Support WebSocket protocol

nkuba opened the pull request:
Here we add support for WebSocket protocol. With this change the client will support all connection types defined by the electrum protocol: `tcp`, `ssl`, `ws` and `wss`.

The WebSocket client initialization expects remote server details to be provided as URL including a scheme and host, e.g. `tcp://electrum.io:50001`.

Examples of client initialization:
```go
// TCP
client, err := NewClient(ctx, "tcp://electrum.io:50001", nil)
// SSL
client, err := NewClient(ctx, "ssl://electrum.io:50002", nil)
// WS
client, err := NewClient(ctx, "ws://electrum.io:50003", nil)
// WSS
client, err := NewClient(ctx, "wss://electrum.io:50004", nil)
```

nkuba committed:
Support WebSocket protocol

Here we add support for WebSocket protocol. With this change the client
will support all connection types defined by the electrum protocol:
`tcp`, `ssl`, `ws` and `wss`.

The WebSocket client initialization expects remote server details to be
provided as URL including a scheme and host, e.g. `tcp://electrum.io:50001`.

nkuba committed:
Expose generic client initialization function

The `NewClient` function should be used to initialize the client.
The function resolves transport protocol from the URL.
It supports all four protocols: `tcp`, `ssl`, `ws` and `wss`.

nkuba commented:
Sample integration tests can be found in https://github.com/keep-network/keep-core/pull/3582.

lukasz-zimnoch commented on electrum/transport_ws.go:
```diff
@@ -0,0 +1,104 @@
+
+// Close closes WebSocket transport.
+func (t *WebSocketTransport) Close() error {
+	// Cleanly close the connection by sending a close message and then
+	// waiting (with timeout) for the server to close the connection.
+	err := t.conn.WriteMessage(websocket.CloseMessage, websocket.FormatCloseMessage(websocket.CloseNormalClosure, ""))
+	if err != nil {
+		log.Printf("%s [error] %s -> close error: %s", time.Now().Format("2006-01-02 15:04:05"), t.conn.RemoteAddr(), err)
+	}
```
Looks like the comment is a bit misleading as we neither wait for the server response nor use a timeout.  Looks like this is taken from https://github.com/gorilla/websocket/blob/76ecc29eff79f0cedf70c530605e486fc32131d1/examples/echo/client.go#L71 where the above is actually done.

According to WS specification (https://datatracker.ietf.org/doc/html/rfc6455#section-7.1.1) the flow for closing the connection should be:
1. Client asks for closing the connection
2. Server closes the connection and notifies the client
3. The client waits for closure notification. If not received after a reasonable amount of time, the client may close the connection on their own

I think we should implement the above pattern here.

nkuba replied:
d5f2ae327a241e525c5e300a85ab364c8324ce67

lukasz-zimnoch commented on electrum/transport_ws.go:
```diff
@@ -0,0 +1,104 @@
+	tlsConfig *tls.Config,
+) (*WebSocketTransport, error) {
+	dialer := websocket.Dialer{
+		TLSClientConfig: tlsConfig,
```
Should we enforce that `tlsConfig` must be `nil` for `ws` connections or that's fine to not bother about it at all? Conversely, shouldn't we enforce it is non-nil for `wss`?

nkuba replied:
I don't think we should enforce `tlsConfig` presence or absence. It's optional and the application using the client should decide if it needs to provide any additional config.

nkuba committed:
Wait for close message response

After sending a close message to the server we need to wait for the
response or force the conncetion closure after a timeout.
